### PR TITLE
networkd: Allow DHCP4 client to set the number to attempt to configure/reconfigure

### PR DESCRIPTION
**Pull request by ssahani**
Otherwise current value is 64 and after 64 it will give up.

Follow up https://github.com/systemd/systemd/pull/12429

### CODE REVIEW

**Comment by yuwata** [comment]
@ssahani I've slightly updated your commit, and added another test for `LinkLocalAddressing=fallback`.

One change from your original is that previously `sd_dhcp_client::attempt` was 2^n, where `n` is the number of attempt. So, calling its upper value as `MaxClientAttempt=` is quite misleading. So, before applying your commit, I made `sd_dhcp_client::attempt` store `n`.

PTAL.

File: test/test-network/systemd-networkd-tests.py [hunk 1]
```diff
@@ -2213,6 +2215,24 @@ def test_dhcp_client_with_ipv4ll_fallback(self):
...
+    def test_dhcp_client_with_ipv4ll_fallback_without_dhcp_server(self):
+        self.copy_unit_to_networkd_unit_path('25-veth.netdev', 'dhcp-server-veth-peer.network',
+                                             'dhcp-client-with-ipv4ll-fallback-without-dhcp-server.network')
+        self.start_networkd(0)
+        self.wait_online(['veth99:degraded', 'veth-peer:routable'])
+
+        output = subprocess.check_output(['ip', 'address', 'show', 'dev', 'veth99']).rstrip().decode('utf-8')
+        print(output)
+
+        output = subprocess.check_output(['ip', '-6', 'address', 'show', 'dev', 'veth99', 'scope', 'global', 'dynamic']).rstrip().decode('utf-8')
```
Comment (ssahani) [inline test/test-network/systemd-networkd-tests.py:2227]: we can look for ```IPv4LL``` address (169.254* )here  that would make more sense.

**Comment by ssahani** [comment]
LGTM to me just one comment.

File: man/systemd.network.xml [hunk 2]
```diff
@@ -1396,6 +1396,15 @@
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term><varname>MaxClientAttempt=</varname></term>
+          <listitem>
+            <para>Specifies how many attempts the DHCPv4 client initiate reconfiguration. Takes a number or
+            <literal>infinity</literal>. Defaults to 6.
+            </para>
```
Comment (keszybz) [inline man/systemd.network.xml]: Let's call this "MaxAttempts=". It certainly should be plural. And "Client" is unneeded, the whole [DHCP] section is about client settings. I think it'd also be nice to add a note about exponential backoff:

"""
Specifies how many times the DHCPv4 client configuration should be attempted. Takes a number or "infinity". Defaults to 6.

The time between retries is increased exponentially, so the network will not be overloaded even if this number if high.
""""

File: src/libsystemd-network/sd-dhcp-client.c [hunk 3]
```diff
@@ -522,6 +523,14 @@ int sd_dhcp_client_set_mtu(sd_dhcp_client *client, uint32_t mtu) {
         return 0;
 }
 
+int sd_dhcp_client_set_max_client_attempt(sd_dhcp_client *client, uint64_t attempt) {
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: No need to repeat "client" here either. When the setting name is changed, this should be adjusted to match.

File: src/network/networkd-network.c [hunk 4]
```diff
@@ -1575,6 +1575,48 @@ int config_parse_section_route_table(
...
+        assert(lvalue);
+        assert(rvalue);
+
+        if (isempty(rvalue)) {
+                network->dhcp_max_client_attempt = 0;
+                return 0;
+        }
+
+        if (streq(rvalue, "infinity"))
+                network->dhcp_max_client_attempt = (uint64_t) -1;
```
Comment (keszybz) [inline src/network/networkd-network.c]: Either use `MAX_CLIENT_ATTEMPT_INFINITY` here, or remove the `MAX_CLIENT_ATTEMPT_INFINITY` define.

File: src/network/networkd-network.c [hunk 5]
```diff
@@ -1575,6 +1575,48 @@ int config_parse_section_route_table(
...
+        else {
+                r = config_parse_uint64(unit, filename, line, section, section_line, lvalue, ltype, rvalue, &k, userdata);
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to parse DHCP maximum client attempt: %s, ignoring.", rvalue);
+                        return 0;
+                }
+
+                network->dhcp_max_client_attempt = k;
```
Comment (keszybz) [inline src/network/networkd-network.c]: The temporary variable is not needed, it's OK to just call `config_parse_uint64(..., &network->dhcp_max_client_attempt, ...)` directly.

File: man/systemd.network.xml (hunk 2 above)
Comment (ssahani) [reply man/systemd.network.xml]: It's actually defaults to ```infinity```. So updated accordingly.

**Comment by ssahani** [comment]
updated @keszbyz . Thanks for review.

File: man/systemd.network.xml (hunk 2 above)
Comment (yuwata) [reply man/systemd.network.xml]: > It's actually defaults to `infinity`. So updated accordingly.

Yeah. Thank you.

**Comment by yuwata** [comment]
@ssahani and @keszybz A forgotten request to the man page is addressed. PTAL.

**Comment by ssahani** [comment]
The test is failing else it's OK.
```
======================================================================
ERROR: test_dhcp_client_with_ipv4ll_fallback_without_dhcp_server (__main__.NetworkdNetworkDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 2222, in test_dhcp_client_with_ipv4ll_fallback_without_dhcp_server
    self.wait_online(['veth99:degraded', 'veth-peer:routable'])
  File "./test/test-network/systemd-networkd-tests.py", line 192, in wait_online
    subprocess.check_call(args)
  File "/usr/lib64/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/lib/systemd/systemd-networkd-wait-online', '--timeout=20s', '--interface=veth99:degraded', '--interface=veth-peer:routable']' returned non-zero exit status 1.

----------------------------------------------------------------------
Ran 79 tests in 1033.880s

```

**Comment by yuwata** [comment]
The test is introduced by this PR, and should not fails. It may be flaky. I'd like to see how often the test fails. Hence, let's restart the CI.

**Comment by yuwata** [comment]
I notice now the test also fails on my local environment. I will rework.

File: src/libsystemd-network/sd-dhcp-client.c [hunk 6]
```diff
@@ -90,7 +88,8 @@ struct sd_dhcp_client {
         uint32_t mtu;
         uint32_t xid;
         usec_t start_time;
-        unsigned attempt;
+        uint64_t attempt;
+        uint64_t max_client_attempt;
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: attempts

Dunno, if we have `attempt`, this should be `max_attempts`, without "client" to match.

File: src/libsystemd-network/sd-dhcp-client.c [hunk 7]
```diff
@@ -522,6 +521,14 @@ int sd_dhcp_client_set_mtu(sd_dhcp_client *client, uint32_t mtu) {
         return 0;
 }
 
+int sd_dhcp_client_set_max_attempt(sd_dhcp_client *client, uint64_t attempt) {
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: attempts

**Comment by yuwata** [comment]
@keszybz Thanks. Updated. PTAL.

File: src/network/networkd-network.c [hunk 8]
```diff
@@ -1575,6 +1575,45 @@ int config_parse_section_route_table(
...
+                return 0;
+        }
+
+        if (streq(rvalue, "infinity")) {
+                network->dhcp_max_attempts = (uint64_t) -1;
+                return 0;
+        }
+
+        r = safe_atou64(rvalue, &network->dhcp_max_attempts);
+        if (r < 0) {
```
Comment (keszybz) [inline src/network/networkd-network.c:1609]: Hmm, 0 is equivalent to infinity. It'd be better to reject 0 explicitly. It seems sloppy to accept 0 here, and then do nothing with it later.

Comment (yuwata) [reply src/network/networkd-network.c:1609]: Thanks. Updated.

**Comment by ssahani** [comment]
This PR is actually tempting me to add a feature like 
dhcp client: use arp to find ip address collisions

https://android.googlesource.com/platform/system/connectivity/dhcp_client/+/efb0b950416a94e649cf3501b812cd9d282093f5

If you remember @yuwata we had a issue DHCP IP configured as static IP then this is the PR we can resolve it.

**Comment by yuwata** [comment]
Yeah, sounds nice! The issue is #11458 and the PR is #11460.

**Pull request merged**